sdks/python/apache_beam/metrics/cells.py

```diff
@@ -0,0 +1,229 @@
+
+
+class CellCommitState(object):
+  """Keeps track of a cell's commit status.
+
+  It's thread-safe.
+  """
+  DIRTY = 0
+  CLEAN = 1
+  COMMITTING = 2
+
+  def __init__(self):
+    self._lock = threading.Lock()
+    self._state = CellCommitState.DIRTY
+
+  @property
+  def state(self):
+    return self._state
+
+  def modified(self):
+    with self._lock:
+      self._state = CellCommitState.DIRTY
+
+  def after_commit(self):
+    with self._lock:
+      if self._state == CellCommitState.COMMITTING:
+        self._state = CellCommitState.CLEAN
+
+  def before_commit(self):
+    with self._lock:
+      if self._state == CellCommitState.CLEAN:
+        return False
+      else:
+        self._state = CellCommitState.COMMITTING
+        return True
+
+
+class MetricCell(object):
+  """Base class of Cell that tracks state of a metric during pipeline execution.
+
+  All subclasses must be thread safe, as these are used in the
+  pipeline runners, and may be subject to parallel/concurrent
+  updates.
+  """
+  def __init__(self):
+    self.commit = CellCommitState()
+    self._lock = threading.Lock()
+
+  def get_cumulative(self):
+    raise NotImplementedError
+
+
+class CounterCell(Counter, MetricCell):
+  """Keeps the state of a counter metric during pipeline execution.
+
+  It's thread safe.
+  """
+  def __init__(self, *args):
+    super(CounterCell, self).__init__(*args)
+    self.value = 0
+
+  def combine(self, other):
+    result = CounterCell()
+    result.inc(self.value + other.value)
+    return result
+
+  def inc(self, n=1):
+    with self._lock:
+      self.value += n
+      self.commit.modified()
+
+  def get_cumulative(self):
+    with self._lock:
+      return self.value
+
+
+class DistributionCell(Distribution, MetricCell):
+  """Keeps the state of a distribution metric during pipeline execution.
+
+  It's thread safe.
+  """
+  def __init__(self, *args):
+    super(DistributionCell, self).__init__(*args)
+    self.data = DistributionData(0, 0, None, None)
+
+  def combine(self, other):
+    result = DistributionCell()
+    result.data = self.data.combine(other.data)
+    return result
+
+  def update(self, value):
+    with self._lock:
+      self.commit.modified()
+      self._update(value)
+
+  def _update(self, value):
+    self.data._count += 1
+    self.data._sum += value
+    self.data._min = (value
+                      if self.data.min is None or self.data.min > value
+                      else self.data.min)
+    self.data._max = (value
+                      if self.data.max is None or self.data.max < value
+                      else self.data.max)
+
+  def get_cumulative(self):
+    with self._lock:
+      return self.data.get_cumulative()
+
+
+class DistributionData(object):
+  """The data structure that holds data about a distribution metric.
+
+  This object is not thread safe, so it's not supposed to be modified
+  by other than the DistributionCell that contains it.
```

Comment:
In the Java version of this code, there are two classes: (1) DistributionData for accumulating partial distributions during pipeline execution and (2) DistributionResult for representing these results as a user API.  For example, this means that the .mean() method should be removed from DistributionData (since it should never be called during accumulation of partial distributions) but included in DistributionResult; correspondingly, the combine() method should be included in DistributionData but not in DistributionResult.  Should we do the same thing here?

CC: @bjchambers for comment.

```java
/**
 * Data describing the the distribution. This should retain enough detail that it can be combined
 * with other {@link DistributionData}.
 *
 * <p>This is kept distinct from {@link DistributionResult} since this may be extended to include
 * data necessary to approximate quantiles, etc. while {@link DistributionResult} would just include
 * the approximate value of those quantiles.
 */
```